spark3/src/main/java/org/apache/iceberg/spark/actions/rewrite/Spark3BinPackStrategy.java

```diff
@@ -0,0 +1,87 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.actions.rewrite;
+
+import java.util.List;
+import java.util.Set;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.actions.BinPackStrategy;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.spark.FileRewriteCoordinator;
+import org.apache.iceberg.spark.FileScanTaskSetManager;
+import org.apache.iceberg.spark.SparkReadOptions;
+import org.apache.iceberg.spark.SparkWriteOptions;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.internal.SQLConf;
+
+public class Spark3BinPackStrategy extends BinPackStrategy {
+  private final Table table;
+  private final SparkSession spark;
+  private final FileScanTaskSetManager manager = FileScanTaskSetManager.get();
+
+  public Spark3BinPackStrategy(Table table, SparkSession spark) {
+    this.table = table;
+    this.spark = spark;
+  }
+
+  @Override
+  public Table table() {
+    return table;
+  }
+
+  @Override
+  public Set<DataFile> rewriteFiles(String groupID, List<FileScanTask> filesToRewrite) {
+    manager.stageTasks(table, groupID, filesToRewrite);
+
+    // Disable Adaptive Query Execution as this may change the output partitioning of our write
+    SparkSession cloneSession = spark.cloneSession();
+    cloneSession.conf().set(SQLConf.ADAPTIVE_EXECUTION_ENABLED().key(), false);
+
+    Dataset<Row> scanDF = cloneSession.read().format("iceberg")
+        .option(SparkReadOptions.FILE_SCAN_TASK_SET_ID, groupID)
+        .option(SparkReadOptions.SPLIT_SIZE, Long.toString(targetFileSize()))
+        .option(SparkReadOptions.FILE_OPEN_COST, "0")
+        .load(table.name());
+
+    // write the packed data into new files where each split becomes a new file
+    FileRewriteCoordinator rewriteCoordinator = FileRewriteCoordinator.get();
+    try {
+      scanDF.write()
+          .format("iceberg")
+          .option(SparkWriteOptions.REWRITTEN_FILE_SCAN_TASK_SET_ID, groupID)
+          .mode("append")
+          .save(table.name());
+    } catch (Exception e) {
+      try {
```

Comment:
Who is responsible for cleaning up resources? It seems we do that in two places: in strategy and in the action. I think we will have to eventually either delegate more to the strategy interface or create a rewriter interface that would be responsible for rewriting, committing, aborting.

Right now both Spark 3 action as well as as Spark 3 bin packing strategy interact with the task set manger and commit coordinator. Ideally, I'd have one just one entity doing that.

Reply:
I think we can ignore this for now and wait for more feedback.

Reply:
Currently the difficulty here is with our commit process, in both cases only the Action knows whether or not it has enough information to do the commit. We could put the commit api into the Strategy as well like how i've setup the action at the moment but that means changing the strategy API again and including more group tracking into the strategy.

For example, in this case the Strategy has the rewriteCoord + manager but other implementations may not require this. A Strategy without a cache like this that just reads and writes files would need an api like

```java
"commit(Files)"
```

While this implementation would need
```java
"commit(groupId)"
```

This is why I had to add in the "groupId" to strategy for rewrite, so we would have some way of keeping
track of the groupID outside of the strategy. If we wanted to make Strategy always responsible for being able to keep track of all files based on "groupId" then I think we would be ok. Then I would do an api like

```java

  // Writes new files from filesToRewrite and associates the result with groupID
  public void rewriteFiles(String groupID, List<FileScanTask> filesToRewrite)
  // Returns added files associated with groupID
  public Set<DataFiles> addedFiles(String groupID)
  // Returns file tasks which were rewritten in groupID
  public Set<DataFiles> rewrittenFiles(String groupID)
  
  // Performs the actual commit operation for the associated groupIDs, removes groupIds when complete
  public void commitGroupIDs(Set<String> groupID)
  // Removes all added files associated with groupID, removes groupIds when complete
  public void abortGroupIDs(Set<String> groupID)
  ```

Reply:
I think we probably would wrap this whole thing into a new api, and make it a transient member of a strategy or something like that? Otherwise I think we have serializability errors ...

Reply:
It's annoying because "rewriteFiles" will be strategy specific and framework specific but all the other methods should just be framework specific